### PR TITLE
[eclipse/xtext-eclipse#1547] enhance ecore2xtext and xtext2ecore

### PR DESCRIPTION
**Pull request by cdietrich**
[eclipse/xtext-eclipse#1547] enhance ecore2xtext generator & xtext2ecore transformer for reserved keywords in enums

Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.xtext.ide/src-gen/org/eclipse/xtext/xtext/ide/contentassist/antlr/internal/InternalXtext.g [hunk 1]
```diff
@@ -9874,9 +9874,9 @@ rule__EnumLiteralDeclaration__EnumLiteralAssignment_0
 	(
 		{ before(grammarAccess.getEnumLiteralDeclarationAccess().getEnumLiteralEEnumLiteralCrossReference_0_0()); }
 		(
-			{ before(grammarAccess.getEnumLiteralDeclarationAccess().getEnumLiteralEEnumLiteralIDTerminalRuleCall_0_0_1()); }
```
Comment (cdietrich) [inline org.eclipse.xtext.xtext.ide/src-gen/org/eclipse/xtext/xtext/ide/contentassist/antlr/internal/InternalXtext.g:9877]: wonder why i dont see any change in ide parser

Comment (szarnekow) [reply org.eclipse.xtext.xtext.ide/src-gen/org/eclipse/xtext/xtext/ide/contentassist/antlr/internal/InternalXtext.g:9877]: How do you generate the code?

Comment (szarnekow) [reply org.eclipse.xtext.xtext.ide/src-gen/org/eclipse/xtext/xtext/ide/contentassist/antlr/internal/InternalXtext.g:9877]: e.g. I use these args in the launch config:
```src/org/eclipse/xtext/xtext/bootstrap/GenerateXtext.mwe2 --param eclipseProjectPath=${project_loc:/org.eclipse.xtext.xtext.ui}```

I suspect that the IDE code was generated to another location.

Comment (szarnekow) [reply org.eclipse.xtext.xtext.ide/src-gen/org/eclipse/xtext/xtext/ide/contentassist/antlr/internal/InternalXtext.g:9877]: Something else: isn't this the IDE parser?

Comment (cdietrich) [reply org.eclipse.xtext.xtext.ide/src-gen/org/eclipse/xtext/xtext/ide/contentassist/antlr/internal/InternalXtext.g:9877]: i just ran GenerateXtext from eclipse

Comment (cdietrich) [reply org.eclipse.xtext.xtext.ide/src-gen/org/eclipse/xtext/xtext/ide/contentassist/antlr/internal/InternalXtext.g:9877]: ahhh it actually is

Comment (szarnekow) [reply org.eclipse.xtext.xtext.ide/src-gen/org/eclipse/xtext/xtext/ide/contentassist/antlr/internal/InternalXtext.g:9877]: :+1:

File: org.eclipse.xtext.tests/xtend-gen/org/eclipse/xtext/xtext/ecoreInference/Xtext2EcoreTransformerTest.java [hunk 2]
```diff
@@ -2091,6 +2095,33 @@ public void testBug296496_02() throws Exception {
...
+  @Test
+  public void testEclipseIssue1547() throws Exception {
+    StringConcatenation _builder = new StringConcatenation();
+    _builder.append("grammar test with org.eclipse.xtext.common.Terminals");
+    _builder.newLine();
+    _builder.append("generate myDsl \'uri\'");
+    _builder.newLine();
+    _builder.append("Model: element=Element;");
+    _builder.newLine();
+    _builder.append("enum Element : ^false=\'false\' | ^true=\'true\';");
```
Comment (szarnekow) [inline org.eclipse.xtext.tests/xtend-gen/org/eclipse/xtext/xtext/ecoreInference/Xtext2EcoreTransformerTest.java:2107]: Quoting shouldn't be necessary anymore given that we call ValidID now.

Comment (szarnekow) [reply org.eclipse.xtext.tests/xtend-gen/org/eclipse/xtext/xtext/ecoreInference/Xtext2EcoreTransformerTest.java:2107]: For other IDs it is necessary though.

Comment (cdietrich) [reply org.eclipse.xtext.tests/xtend-gen/org/eclipse/xtext/xtext/ecoreInference/Xtext2EcoreTransformerTest.java:2107]: ?

Comment (cdietrich) [reply org.eclipse.xtext.tests/xtend-gen/org/eclipse/xtext/xtext/ecoreInference/Xtext2EcoreTransformerTest.java:2107]: ok will test both.

File: org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/ecore2xtext/Ecore2XtextExtensions.xtend [hunk 3]
```diff
@@ -172,7 +172,9 @@ class Ecore2XtextExtensions {
 			'as',
 			'current',
 			'fragment',
-			'EOF'
+			'EOF',
+			'true',
+			'false'
```
Comment (szarnekow) [inline org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/ecore2xtext/Ecore2XtextExtensions.xtend:177]: I see that isXtextKeyword should return true for `true`and `false`, but it doesn't really help given how this method is used.

**Review by szarnekow — APPROVED** [review]

File: org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/ecore2xtext/Ecore2XtextExtensions.xtend (hunk 3 above)
Comment (cdietrich) [reply org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/ecore2xtext/Ecore2XtextExtensions.xtend:177]: see org.eclipse.xtext.xtext.wizard.ecore2xtext.Ecore2XtextExtensions.quoteIfNeccesary(String)

Comment (cdietrich) [reply org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/ecore2xtext/Ecore2XtextExtensions.xtend:177]: looks like we dont have a consistent concept for when to use ^ there

Comment (cdietrich) [reply org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/ecore2xtext/Ecore2XtextExtensions.xtend:177]: (maybe for some backwards compatibility, so i would keep it for consistency

Comment (cdietrich) [reply org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/ecore2xtext/Ecore2XtextExtensions.xtend:177]: so as we use validID everywhere we also could drop it here

Comment (szarnekow) [reply org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/ecore2xtext/Ecore2XtextExtensions.xtend:177]: > so as we use validID everywhere we also could drop it here

:+1:

Comment (cdietrich) [reply org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/ecore2xtext/Ecore2XtextExtensions.xtend]: done

**Pull request merged**